### DISCUSSION TITLE
Cassini data

### DISCUSSION THREAD

**Discussion by cmoissar** [SciQLop/speasy#99] [Q&A] [answered]
This may not be the right place to ask this question.

I am looking for Cassini data, which can be found here:

spz.inventories.tree.amda.Parameters.Cassini

Then I admit to being quite lost.
Do you have examples or a link to helpful documentation?

Thank you!

**Comment by brenard-irap** [comment]
Hi Clément,

Yes, you have access to all data available in AMDA: Ephemeris, CAPS (moments and spectra), MAG (orbital + cruise + Jupyter flyby), MIMI-LEMMS and RPWS (parameters derived from Langmuir Probe + skr).

For example:

```python
import speasy as spz
amda_tree = spz.inventories.data_tree.amda
cass_b_rtn = spz.get_data(amda_tree.Parameters.Cassini.MAG.orbit_saturn.cass_mag_orb1.cass_b_rtn1s, '2008-01-01', '2008-01-02')
print(cass_b_rtn.to_dataframe())
```

Which data are you looking for?

**Reply by cmoissar** [reply]
Hi Benjamin, thanks a lot for this.

I am looking for the ``common'' plasma parameters: (n, T, v, B) and the position of the satellite (+ which frame of reference is it in?).

I am very new to any data that's far from Earth, so there's a bit of a learning curve!

Thank you very much for your help.

**Reply by jeandet** [reply]
@vgenot I guess you may be able to answer better than us 😅.

**Comment by vgenot** [comment]
salut Clément,
for moments look into CAPS but I know they must taken with care (when they're available !). The magnetic field dataset is complete. All from PDS. As for the positions you'll find them in ephemeris in KSO, KSM, KRTP or IAU lon/lat depending on your usage.
You can always come back to AMDA to browse the data tree :-) and get some info/doc

**Reply by cmoissar** [reply]
Merci Vincent!

**Comment by Luke-Bicknell** [comment]
Hello

Clement (cmoissar) is my project supervisor for my BSc physics project and I am interested in the Cassini mission and gathering data from Cassini. The information above is helpful, but I am on AMDA and wondered where best to start with the spacecraft's position for example.

![image](https://github.com/SciQLop/speasy/assets/147398510/e05dde97-4a66-419c-8c9f-2da356a2de80)

The link above is an image of the data tree, and I didn't know which folder was best to start.

Is there a way of seeing what the exact values of local minima/maxima are on a graph? For example, Cassini's closest approach to Saturn.

Sorry for the overload! Many thanks

**Reply by brenard-irap** [reply]
Hello,

**The link above is an image of the data tree, and I didn't know which folder was best to start.**

If you are interested by Cassini's closest approach to Saturn, you should probably use a dataset defined in "orbit saturn".
There are 3 different datasets for 3 different sampling time: 1 min, 1 sec or 5 min (the most suitable depends on your study).

**Is there a way of seeing what the exact values of local minima/maxima are on a graph? For example, Cassini's closest approach to Saturn.**

You can use the "Statistics module" (not directly from the Plot Module).

![statistics](https://github.com/SciQLop/speasy/assets/77481833/daee3dfe-1be4-4ac6-a790-f17f38862a68)

![statistics_result](https://github.com/SciQLop/speasy/assets/77481833/c4f44246-fa48-48b2-aeb4-027148983f56)

You can take a look to this publication: https://doi.org/10.1016/j.pss.2021.105214 to have more information on AMDA with some use cases.
If you have more questions/feedbacks on AMDA, you received a contact email with your AMDA registration (this repository is dedicated to Speasy project).

But, you can do the same stuff with Speasy!
```python

import speasy as spz

amda_tree = spz.inventories.data_tree.amda

# '5 min' dataset:
cassini_orbit_dataset = amda_tree.Parameters.Cassini.Ephemeris__Cassini.orbit_saturn.cass_orb_saturn

cass_r_sat = spz.get_data(cassini_orbit_dataset.cass_r_sat,'2010-01-01', '2011-01-01')

print(cass_r_sat.to_dataframe().describe())
```

The result will be:

```
          cass_r_sat
count  105107.000000
mean       31.526586
std        11.397203
min         2.574535
25%        24.254516
50%        34.832123
75%        40.224302
max        49.170737
```

**Reply by Luke-Bicknell** [reply]
@brenard-irap Thank you so so much for this, that's helped a lot!

Have a great day :)

**Comment by Luke-Bicknell** [comment]
Good afternoon all

I just wanted to ask another question regarding speasy

I am slightly confused as to what the following refers to

![image](https://github.com/SciQLop/speasy/assets/147398510/702f7ebb-0493-4b1b-9059-6eb10ae7de8b)

![image](https://github.com/SciQLop/speasy/assets/147398510/24d8c5d9-ad57-4678-9b2d-a83d15559e2e)

If someone would briefly explain that would be a massive help! After some plotting I get a 3D plot, but I'm not sure how the language makes it so?

Many thanks,
Luke

**Reply by cmoissar** [reply]
Hi Luke, this is not so much speasy specific as it is a piece of code I wrote using speasy.

`ace_mag` contains all the information about the magnetic field measured by ace during a certain time period that I selected earlier in the code.

`ace_mag.time` will give you the time at which every "value" was measured.
`ace_mag.values` will give you the magnetic field (as a vector) at these times.
`ace_mag.unit` should be nano Teslas.

`start_time = time.time()`  has nothing to do with speasy, and gets the clock time of your computer.
"IMF" means interplanetary magnetic field.
`Bx = ace_mag.filter_colums(["imf[0]").values` will give you an array containing the values of Bx at each of the above times.
By and Bz are the same.
`print("--- %s seconds ---" % (time.time() - start_time))` prints the time it took to perform all the operations since you first defined `start_time`.

You can forget about the `time.time()` lines. I was simply trying to compare how long it would take to define the magnetic field as arrays using two different methods. The other one, which used "list comprehensions" (we can talk about this next Wednesday) is probably not in that piece of code anymore because it was much slower that the one above.

So at this point, you have arrays containing Bx, By, Bz.

The next few lines use... "list comprehensions" (probably not a great idea) to define arrays that contain the coordinates of the ACE satellite over time.


I can see how you would plot a 3D graph giving the trajectory of ACE based on these three lines. I am not sure what other 3D graph you may be referring to.

Was this the answer you were looking for?

**Reply by Luke-Bicknell** [reply]
Hi Clement. Thank you for the speedy reply!

I see now the 3D plot mapped the different positions of the Earth, moon and ACE satellite I think the other info just contains details about the plot aesthetics. It is a very cool plot!

This helps a lot thank you. I've started also to get the angles of Enceladus so hopefully I'm still on-track.

Many thanks

**Comment by Luke-Bicknell** [comment]
Does anyone know how to access Enceladus data? I can't seem to access it through the AMDA tree using speasy.

I should say I can manually find it on the AMDA database, but not using speasy indirectly.

**Reply by jeandet** [reply]
@Luke-Bicknell If you mean these products:
![image](https://github.com/SciQLop/speasy/assets/7570249/35e2a501-1d1d-4af2-bbbd-1183fbdb7dbe)
They are under this node `spz.inventories.tree.amda.Parameters.Cassini.Ephemeris__Saturn_Moons___special.Enceladus` in speasy. Speasy tries to keep the same names than AMDA as long as they are "Python compatible" or replace spaces and others forbidden characters by `_`.

**Reply by Luke-Bicknell** [reply, marked as answer]
Thank you! @jeandet I must have typed something incorrectly before

**Comment by Luke-Bicknell** [comment]
Hi all,

In KRTP coordinates, I have the radial and theta positions of Cassini and wanted to pass them into the radial component of the magnetic dipole equation below:
![image](https://github.com/SciQLop/speasy/assets/147398510/70fc5e30-b85c-4508-9e27-172e0fbdcb53)

If I have an array of radial and theta values respectively, how can I pass these simultaneously into the B_r function?

I am a novice at coding, so below is my initial attempt for reference:
![image](https://github.com/SciQLop/speasy/assets/147398510/b754e7d5-a2b2-4e07-9ad0-c7c60e2f8ade)

Many thanks,
Luke

**Comment by jeandet** [comment]
@Luke-Bicknell, I don't get question exactly. I'm also not sure about your for loop. I believe you want more something like this:

```Python
for r,theta in zip(r_cass_meters, theta_cass):
    ...
```

**Reply by Luke-Bicknell** [reply]
@jeandet Thank you so much for the reply, much appreciated.

Yes so to put it more plainly I wanted to plug r and theta into the following equation/function:
![image](https://github.com/SciQLop/speasy/assets/147398510/7f755ec2-a96a-44f8-8fb2-0c3dda59664b)

**Reply by jeandet** [reply]
I would say your code should work as long as you use numpy `cos` function and **r**, **theta** are numpy arrays.
